Complete textures / text color coverage for heracles themes

sisby-folk opened the pull request:
Closes #134 (vanilla style buttons can be handled by another Issue/PR)

Main:

![image](https://github.com/terrarium-earth/Heracles/assets/55819817/b5ec65b9-1468-4a1a-8f9a-9747607eb2a4)

PR: 

![javaw_4bo46h5L25](https://github.com/terrarium-earth/Heracles/assets/55819817/5ef79bea-9d88-4104-9446-56ab008eb8c4)

PR + Respack:

![javaw_7K5Y88VZkV](https://github.com/terrarium-earth/Heracles/assets/55819817/f8420eab-e6f0-4df8-b2f1-6b73676a271d)

sisby-folk committed:
fix bugs with selectablerewardwidget + theme
Prevents claim buttons from appearing on select options
Prevents no rewards from being claimed
make selected widget an overlayed texture.
(a bg variant would be better, but complex)

sisby-folk committed:
move widget separator rendering

sisby-folk committed:
extract more common widget backgrounds

sisby-folk committed:
texturable widget bg, summary, and list headers

sisby-folk committed:
texturable task/reward status headers, progress bar

sisby-folk committed:
texturable AddDisplayWidget w/ glyph, move bars

sisby-folk commented:
Realized the texture for "selected reward widget" made sense alongside the widget background, so I've added that.

Also fixed claim buttons appearing on those widgets, and being able to claim 0 rewards.

![image](https://github.com/terrarium-earth/Heracles/assets/55819817/96ef1d03-59a1-43d5-9fb5-0a3912a22e57)

sisby-folk committed:
Merge branch '1.20.x' into 1.20.x_slices

sisby-folk committed:
Themeable groups list backgrounds

sisby-folk commented:
Add theming for the groups list

https://github.com/terrarium-earth/Heracles/assets/55819817/4f07141e-17c3-4024-8914-fca364e406e8

sisby-folk committed:
Themeable quests screen scrollbars

sisby-folk commented:
Added theming for quests view scrollbars

https://github.com/terrarium-earth/Heracles/assets/55819817/91356312-c029-4f32-9742-f3e5ef9b3a24

sisby-folk committed:
Themeable modal background

sisby-folk commented:
Added theming for the background for modals (modals themselves were already themeable)

![javaw_DXDtk6xmNL](https://github.com/terrarium-earth/Heracles/assets/55819817/4d5de851-bf7e-4982-8991-d5c0dc55f2b7)
![javaw_t8u4Ig3M45](https://github.com/terrarium-earth/Heracles/assets/55819817/81bf3b9a-8363-4f72-b433-998b82aa3c71)

Also extended the background to cover the header - this avoids strange artifacts with themes that don't have a perfectly square header bar (see above)

sisby-folk committed:
Themeable quest highlight overlay

sisby-folk commented:
Added themeable quest highlight overlays.

![javaw_HLhkI1SNpC](https://github.com/terrarium-earth/Heracles/assets/55819817/e4f21f55-cae8-41ec-9dbd-0732c609db4d)

![javaw_Tho1jrJxqa](https://github.com/terrarium-earth/Heracles/assets/55819817/de445a3a-13a3-47f2-9ce5-d94018cc381a)

These are added to the background textures themselves, so they can match the shape.

sisby-folk committed:
Themeable pinned quest

sisby-folk commented:
Added themeable pinned quests + screen

![javaw_rkmqLX8tBz](https://github.com/terrarium-earth/Heracles/assets/55819817/40d2c0f6-a8f3-4d19-83b9-396348ef79bb)

![javaw_TgvFatCgR6](https://github.com/terrarium-earth/Heracles/assets/55819817/ba952764-a2e3-4a58-b9f3-ffa668c07ba5)

minor uplift in the fake popup on default as well.

sisby-folk committed:
Add fourth status to headings

sisby-folk committed:
Add color constants and loader

sisby-folk committed:
shorten summary keys while i'm here

sisby-folk committed:
Apply theme colors, other minor fixes
Move quest tutorial toast into the toasts folder
Fix the typo in PlacholderEditBox
Use widgetUtils.drawProgressBar more
Remove a couple of stdouts

sisby-folk committed:
Use translatable keys and constant components where possible

sisby-folk committed:
Merge pull request #1 from sisby-folk/1.20.x_slices_theming

Text colors and translatables

sisby-folk commented:
Alright, considering there's no activity here and the PR is going to be uncomfortably large no matter what, we've merged in our text recoloring + translatableifying PR as well.

![image](https://github.com/terrarium-earth/Heracles/assets/55819817/4f0a1320-a707-4f8b-a38b-0f520f1103e1)

![image](https://github.com/terrarium-earth/Heracles/assets/55819817/45136d3d-657c-462c-80d7-24db3c25467e)

![image](https://github.com/terrarium-earth/Heracles/assets/55819817/d5911b5a-43d8-498e-848b-0f81e2fb400d)

ThatGravyBoat commented:
So ive taken a look at the color part of the PR and found a few issues and concerns ive made a PR to fix them https://github.com/sisby-folk/Heracles/pull/2 if you can take a look that would be great!

sisby-folk committed:
Theme vanilla-style buttons (sorry, lol)

sisby-folk committed:
Add click sounds to fake buttons
this one was driving me nuts

ThatGravyBoat committed:
Some changes to color theming (#2)

* Change to using getters and codecs for theme colors

* Fix merge conflicts

sisby-folk commented:
We don't have any plans to tack anything further on this PR btw - merge as soon as you're ready.

ThatGravyBoat commented:
ill review in about an hour when if I have time if not tomorrow

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/client/widgets/buttons/ThemedButton.java:
```diff
@@ -0,0 +1,92 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public interface ThemedButton {
```
I would perfer if this was a class that extended buttons and overrode the methods instead of mixing into button

sleepingdragoninn replied:
It mixes into AbstractButton, and - We tried this at first, but solving the problem this way got very complicated very fast. Consider how I might implement the theme onto `SelectableImageButton` (extends ImageButton extends Button) vs `SelectableTabButton` (extends AbstractButton) - I'd need like four classes and a swath of copied vanilla code that's guaranteed to break - the reimplementation of this in non-mixin in 1.20.2 would be a lot easier than having to redo it for all that I think.

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/client/widgets/modals/SelectRewardsModal.java:
```diff
@@ -30,12 +39,13 @@ public class SelectRewardsModal extends BaseModal {
         super(screenWidth, screenHeight, (int) (screenWidth * 0.75f), (int) (screenHeight * 0.8f));
 
-        addChild(Button.builder(ConstantComponents.Rewards.CLAIM_REWARD, b -> {
+        claimButton = ThemedButton.builder(ConstantComponents.Rewards.CLAIM_REWARD, b -> {
```
addChild returns the type please combine these back to 1 line.

sleepingdragoninn replied:
done in 10efdf3

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/client/handlers/QuestTutorial.java:
```diff
@@ -59,41 +54,15 @@ public static void load(Path path) {
                     FileUtils.write(displayFile, tutorialText, StandardCharsets.UTF_8);
                 } catch (Exception e) {
-                    e.printStackTrace();
+                    Heracles.LOGGER.error("[Heracles] Failed to write tutorial", e);
```
Prefixing is not required as with logger it will print as `[00:00:00] [Render Thread/ERROR] (Heracles) ${message}`

sleepingdragoninn replied:
Isn't that only true in a dev environment, not a prod log file?

ThatGravyBoat replied:
no it also shows in prod, in prod it actually shows as `[00:00:00] [Render thread/INFO] [ea.te.he.Heracles/]: ${message}`

sleepingdragoninn replied:
(latest.log)

![image](https://github.com/terrarium-earth/Heracles/assets/55819817/dce761c9-5d72-4254-9165-e1dce42abaec)

sleepingdragoninn replied:
I've gone ahead and made this change consistent in d0f5138 - i figure this will make it easier to go one way or the other.

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/api/rewards/client/defaults/XpRewardWidget.java:
```diff
@@ -44,6 +45,11 @@ public boolean canClaim() {
     }
 
+    @Override
+    public boolean isInteractive() {
```
name the fields `isInteractive` to remove the need for overriding this method in almost every widget

sleepingdragoninn replied:
(forgets how records work) didn't know that trick! - done in 10efdf3

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/mixins/client/AbstractButtonMixin.java:
```diff
@@ -0,0 +1,34 @@
+import org.spongepowered.asm.mixin.injection.invoke.arg.Args;
+
+@Mixin(AbstractButton.class)
+public abstract class AbstractButtonMixin extends AbstractWidget {
```
Read comment in ThemeButton

sisby-folk committed:
fluent addchild and overriding reward records

sisby-folk commented:
Heya @ThatGravyBoat - I've got some further stuff on the colours that I think need looking at but I can't go about.

I've readded the json default pack-in json file, but it needs to be fixed to the new keys and value formats if they've changed (#FFFFFF format would probably be best) - I think we've gotta include it to be in-line with the other respackable stuff in terms of how intuitive it is to replace. 

I also think that considering there's a fancier system for loading the json now, it's worth considering having the defaults stored in json purely instead of in code - this would prevent accidentally adding it to one but not the other. However it's probably more critical to ensure that respacks can partially override text colours while falling back on defaults (this happens as-is, right?) - otherwise Heracles updates could fail on old respacks.

ThatGravyBoat commented:
There is no need for the JSON as if the data is missing in the JSON it will default back to the value that is in code even if its only 1 line, this also works for old resourcepacks as it will default.

ThatGravyBoat commented:
pack developers would only need to add the elements they need not all, so if they only change the group name they only need to add group data the rest gets used from code

sisby-folk commented:
Ah okay that's good - feel free to remove the json then. Would you be able to send us through the new format using the colours from this pack? 
[Heracles_Paper__Ribbon.zip](https://github.com/terrarium-earth/Heracles/files/13365388/Heracles_Paper__Ribbon.zip)

sisby-folk committed:
don't use snake case for codec, hearcles -> heracles
The mod uses snake case in codecs 3 other times, but camelCase 40 other times - camelcase wins.

sisby-folk commented:
Ignore me, got around to it - here's the sample pack.

[[Heracles] Paper & Ribbon.zip](https://github.com/terrarium-earth/Heracles/files/13370795/Heracles.Paper.Ribbon.zip)

Theme json is removed as well so this is ready for re-review.

sisby-folk committed:
remove "message" from text color names

ThatGravyBoat commented on common/src/main/java/earth/terrarium/heracles/mixins/client/AbstractButtonMixin.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @ModifyArgs(method = "renderWidget", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/GuiGraphics;blitNineSliced(Lnet/minecraft/resources/ResourceLocation;IIIIIIIIII)V"))
+    public void applyThemedTexture(Args args) {
```
prefix your mixin methods with `heracles$`.

sleepingdragoninn replied:
added!

sisby-folk committed:
prefix mixins

sisby-folk committed:
prefix all logging, use debug logger for translation dumping

ThatGravyBoat committed:
Undo prefixing and add PlatformLogger

sisby-folk committed:
protect heracles$applyThemedTexture against redirects

sisby-folk committed:
add hint comment for the mixin issue

sisby-folk committed:
Merge branch '1.20.x' into 1.20.x_slices
